File: test/gallformers/phenology/seasonal_clock_test.exs
defmodule Gallformers.Phenology.SeasonalClockTest do
  use ExUnit.Case, async: true
  alias Gallformers.Phenology.SeasonalClock, as: Clock

  test "client reference contains the exact anchors used by predictions" do
    assert length(Clock.reference()) == 121

    for [lat, spring, autumn] <- Clock.reference() do
      assert Clock.anchors(lat) == {spring, autumn}
    end
  end

  test "selection transfers calendar-day edges, wraps winter and ignores unsupported points" do
    for day <- [1, 90, 180, 270, 355, 366], lat <- [25, 37.5, 55] do
      {:ok, {low, high} = window} = Clock.window(day, lat, 10)
      assert_in_delta Clock.inverse(low, lat), day - 10, 1.0e-8
      assert_in_delta Clock.inverse(high, lat), day + 10, 1.0e-8

      for other <- [25, 40, 55] do
        projected = Clock.coordinate(day, lat) |> Clock.inverse(other)
        wrapped = projected - floor((projected - 1) / 365) * 365
        assert Clock.in_window?(wrapped, other, window) == true
      end

      refute Clock.in_window?(day, nil, window)
      refute Clock.in_window?(day, 24.9, window)
    end

    {:ok, winter} = Clock.window(355, 40, 15)
    assert Clock.in_window?(5, 40, winter) == true
    refute Clock.in_window?(6, 40, winter)
    {:ok, exact} = Clock.window(120, 40, 0)
    assert Clock.in_window?(120, 40, exact) == true
    refute Clock.in_window?(121, 40, exact)
    {:ok, whole_year} = Clock.window(180, 40, 183)
    for day <- 1..366, do: assert(Clock.in_window?(day, 40, whole_year) == true)
    assert Clock.window(100, 60, 10) == :error
    assert Clock.window(100, 40, -1) == :error
    assert Clock.window(100, 40, 184) == :error
    assert Clock.window(nil, 40, 10) == :error
  end

  test "finite latitude support and all interpolated landmarks are ordered" do
    refute Clock.supported_latitude?(nil)
    refute Clock.supported_latitude?(-30)
    refute Clock.supported_latitude?(55.01)

    for i <- 250..550 do
      lat = i / 10
      assert Clock.supported_latitude?(lat) == true
      {spring, autumn} = Clock.anchors(lat)
      assert spring < autumn and autumn < spring + 365

      for day <- [0, 60, 180, 300, 370] do
        assert_in_delta Clock.inverse(Clock.coordinate(day, lat), lat), day, 1.0e-8
      end
    end
  end
end

File: assets/js/hooks/seasonal_clock.test.js
import {describe, test, expect} from 'vitest'
import {readFileSync} from 'node:fs'
import {URL as NodeURL} from 'node:url'
import {createClock, inWindow} from './seasonal_clock'
import {applySeasonalLandmark} from './phenology_state'

const rows = readFileSync(new NodeURL('../../../priv/phenology/seasonal_landmarks.csv', import.meta.url), 'utf8')
  .trim().split('\n').slice(1).map(row => row.split(',').slice(0, 3).map(Number))
const clock = createClock(rows)

describe('shared seasonal-landmark clock', () => {
  test('reference landmarks and continuous forward/inverse coordinates agree', () => {
    for (const [lat, spring, autumn] of rows) {
      expect(clock.coordinate(spring, lat)).toBeCloseTo(0, 8)
      expect(clock.coordinate(autumn, lat)).toBeCloseTo(182.5, 8)
    }
    for (const lat of [25, 30.13, 40, 50.9, 55]) {
      for (const day of [-180, 1, 90, 180, 270, 365, 550]) {
        const phase = clock.coordinate(day, lat)
        expect(clock.inverse(phase, lat)).toBeCloseTo(day, 8)
        expect(clock.coordinate(day + 365, lat)).toBeCloseTo(phase + 365, 8)
      }
    }
  })

  test('reference days define both edges and transfer across supported latitudes', () => {
    for (const day of [1, 90, 180, 270, 355, 366]) {
      for (const lat of [25, 37.5, 55]) {
        const window = clock.window(day, lat, 10)
        expect(clock.inverse(window.low, lat)).toBeCloseTo(day - 10, 8)
        expect(clock.inverse(window.high, lat)).toBeCloseTo(day + 10, 8)
        expect(inWindow(window.low, window)).toBe(true)
        expect(inWindow(window.high, window)).toBe(true)
        expect(inWindow(window.low - 0.001, window)).toBe(false)
        expect(inWindow(window.high + 0.001, window)).toBe(false)
        for (const other of [25, 40, 55]) {
          const projected = clock.inverse(clock.coordinate(day, lat), other)
          const wrapped = projected - Math.floor((projected - 1) / 365) * 365
          const point = {doy: wrapped, lat: other, seasind: null}
          expect(applySeasonalLandmark([point], {clock, window})).toEqual([point])
        }
      }
    }
  })

  test('winter, exact dates, full-year windows and invalid inputs are explicit', () => {
    const winter = clock.window(355, 40, 15)
    expect(inWindow(clock.coordinate(5, 40), winter)).toBe(true)
    expect(inWindow(clock.coordinate(6, 40), winter)).toBe(false)
    const exact = clock.window(120, 40, 0)
    expect(inWindow(clock.coordinate(120, 40), exact)).toBe(true)
    expect(inWindow(clock.coordinate(121, 40), exact)).toBe(false)
    const wholeYear = clock.window(180, 40, 183)
    for (let day = 1; day <= 366; day++) expect(inWindow(clock.coordinate(day, 40), wholeYear)).toBe(true)
    for (const args of [[0, 40, 10], [367, 40, 10], [null, 40, 10], [100, 60, 10],
      [100, 40, null], [100, 40, -1], [100, 40, 184], [100, 40, Infinity]]) {
      expect(clock.window(...args)).toBeNull()
    }
    expect(applySeasonalLandmark([{doy: 120, lat: 20}], {clock, window: wholeYear})).toEqual([])
    expect(applySeasonalLandmark([{doy: 120, lat: 40}], {clock, window: null})).toEqual([])
    expect(() => createClock([])).toThrow('Invalid seasonal-landmark reference')
  })
})
